simulator/main.rs

```diff
@@ -67,6 +116,343 @@ enum Value {
+    Eq(String, Value),
+    Gt(String, Value),
+    Lt(String, Value),
+}
+
+enum Query {
+    Create { table: Table },
+    Select { table: String, predicate: Predicate },
+    Insert { table: String, values: Vec<Value> },
+    Delete { table: String, predicate: Predicate },
+}
+
+impl ArbitraryFrom<Table> for Query {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Table) -> Self {
+        match rng.gen_range(0..=200) {
+            0 => Query::Create {
+                table: Table::arbitrary(rng),
+            },
+            1..=100 => Query::Select {
+                table: t.name.clone(),
+                predicate: Predicate::arbitrary_from(rng, t),
+            },
+            101..=200 => Query::Insert {
+                table: t.name.clone(),
+                values: t
+                    .columns
+                    .iter()
+                    .map(|c| Value::arbitrary_from(rng, &c.column_type))
+                    .collect(),
+            },
+            201..=300 => Query::Delete {
+                table: t.name.clone(),
+                predicate: Predicate::arbitrary_from(rng, t),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+struct CompoundPredicate(Predicate);
+struct SimplePredicate(Predicate);
+
+impl ArbitraryFrom<(&Table, bool)> for SimplePredicate {
+    fn arbitrary_from<R: Rng>(rng: &mut R, (t, b): &(&Table, bool)) -> Self {
+        // Pick a random column
+        let column_index = rng.gen_range(0..t.columns.len());
+        let column = &t.columns[column_index];
+        let column_values = t.rows.iter().map(|r| &r[column_index]).collect::<Vec<_>>();
+        // Pick an operator
+        let operator = match rng.gen_range(0..3) {
+            0 => {
+                if *b {
+                    Predicate::Eq(
+                        column.name.clone(),
+                        Value::arbitrary_from(rng, &column_values),
+                    )
+                } else {
+                    Predicate::Eq(
+                        column.name.clone(),
+                        Value::arbitrary_from(rng, &column.column_type),
+                    )
+                }
+            }
+            1 => Predicate::Gt(
+                column.name.clone(),
+                match b {
+                    true => GTValue::arbitrary_from(rng, &column_values).0,
+                    false => LTValue::arbitrary_from(rng, &column_values).0,
+                },
+            ),
+            2 => Predicate::Lt(
+                column.name.clone(),
+                match b {
+                    true => LTValue::arbitrary_from(rng, &column_values).0,
+                    false => GTValue::arbitrary_from(rng, &column_values).0,
+                },
+            ),
+            _ => unreachable!(),
+        };
+
+        SimplePredicate(operator)
+    }
+}
+
+impl ArbitraryFrom<(&Table, bool)> for CompoundPredicate {
+    fn arbitrary_from<R: Rng>(rng: &mut R, (t, b): &(&Table, bool)) -> Self {
```

Comment:
please try to make variables more meaningful I don't know what b is here.

Reply:
yeah i agree with this, let's use longer more descriptive variable names

Reply:
I've changed all the names to more descriptive versions.